compiler/noirc_evaluator/src/ssa/opt/die/prune_dead_parameters.rs

```diff
@@ -0,0 +1,250 @@
+
+            // Create the list of new params for updating the block with unused parameters
+            // as well as an indexed list of the removed parameters to update each predecessor's terminator argument list.
+            let mut keep_list = Vec::with_capacity(old_params.len());
+            let mut new_params = Vec::with_capacity(old_params.len());
+            let unused_set = unused_params.iter().copied().collect::<HashSet<_>>();
+            for param in old_params {
+                let keep = !unused_set.contains(&param);
+                keep_list.push(keep);
+                if keep {
+                    new_params.push(param);
+                }
+            }
+
+            self.dfg[block].set_parameters(new_params);
+
+            // Update the predecessor argument list to match the new parameter list
+            let predecessors = cfg.predecessors(block);
+            for pred in predecessors {
+                let terminator = self.dfg[pred].unwrap_terminator_mut();
+
+                match terminator {
+                    TerminatorInstruction::JmpIf { .. } => {
+                        // No terminator arguments in a JmpIf, so we do nothing here
+                    }
+                    TerminatorInstruction::Jmp { destination, arguments, .. } => {
+                        // Cannot place this guard on the pattern as we are matching by reference
+                        if *destination == block {
+                            let new_args = arguments
+                                .iter()
+                                .zip(keep_list.iter())
+                                .filter_map(|(arg, &keep)| if keep { Some(*arg) } else { None })
+                                .collect();
+                            *arguments = new_args;
+                        }
+                    }
+                    TerminatorInstruction::Return { .. } => {
+                        unreachable!("ICE: A return block should not be a predecessor");
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::assert_ssa_snapshot;
+
+    use crate::ssa::Ssa;
+
+    #[test]
+    fn prune_unused_block_params() {
+        let src = r#"
+        brillig(inline) fn test f0 {
+          b0():
+            jmp b1(Field 1, Field 2, Field 3)
+          b1(v0: Field, v1: Field, v2: Field):
+            return v1
+        }"#;
+
+        let mut ssa = Ssa::from_str(src).unwrap();
+        // DIE is necessary to fetch the block parameters liveness information
+        ssa = ssa.dead_instruction_elimination();
+        ssa = ssa.prune_dead_parameters();
+
+        assert_all_unused_parameters_cleared(&ssa);
+
+        assert_ssa_snapshot!(ssa, @r#"
+        brillig(inline) fn test f0 {
+          b0():
+            jmp b1(Field 2)
+          b1(v0: Field):
+            return v0
+        }"#);
+    }
+
+    #[test]
+    fn prune_unused_block_params_multiple_predecessors() {
```

Comment:
Can you add a larger test which shows a few unused results from e.g. the merge of a few `if`s?

Reply:
done